### DISCUSSION TITLE
Media Messages : ImageMessage, VideoMessage, DocumentMessage ...

### DISCUSSION THREAD

**Discussion by mamur-rezeki** [tulir/whatsmeow#260] [WhatsApp protocol Q&A] [answered]
## ImageMessage

```go

  if up, err := client.Upload(context.Background(), bytedata, whatsmeow.MediaImage); err != nil {
    return nil, err
  } else {

    var message = &waProto.ImageMessage{
      Url:           &up.URL,
      Mimetype:      proto.String(mimetype),
      Caption:       proto.String("Caption"),
      FileSha256:    up.FileSHA256,
      FileEncSha256: up.FileEncSHA256,
      FileLength:    &up.FileLength,
      MediaKey:      up.MediaKey,
      DirectPath:    &up.DirectPath,
    }
  }

```

**Comment by arthurbaquit** [comment, marked as answer]
I think for sending documents you need to use `waProto.DocumentMessage{...`

**Reply by mamur-rezeki** [reply]
This discussion I've write for templates because so many ask about message protocol, sorry for marking it as a question.

**Comment by mamur-rezeki** [comment]
## Circular Video Message

Circular Video Message is based on `VideoMessage` that put in `PtvMessage` field:

```go
var msg = &waProto.Message{
    PvtMessage: &waProto.VideoMessage{
         ....
    }
}
```

**Reply by Ased2235** [reply]
What is Circular video msg?

**Reply by mamur-rezeki** [reply]
Circular Video Message is a beta testing Whatsapp video message.

_example image from reddit._
![images](https://github.com/tulir/whatsmeow/assets/107448645/4ea3094b-0167-405e-b93d-4552908be492)

_this from Whatsapp beta_
![Screenshot_20230627-082240_DualMeta](https://github.com/tulir/whatsmeow/assets/107448645/91748795-11dd-48b2-8a72-5b33f2b0065c)
